File: win-prog/ProgWinVB/Dialog Boxes/BetterFontAndColorDialogs/BetterFontAndColorDialogs.vb
'----------------------------------------------------------
' BetterFontAndColorDialogs.vb (c) 2002 by Charles Petzold
'----------------------------------------------------------
Imports System
Imports System.Drawing
Imports System.Windows.Forms

Class BetterFontAndColorDialogs
    Inherits Form

    Protected clrdlg As ColorDialog = New ColorDialog()

    Shared Sub Main()
        Application.Run(New BetterFontAndColorDialogs())
    End Sub

    Sub New()
        Text = "Better Font and Color Dialogs"

        Menu = New MainMenu()
        Menu.MenuItems.Add("&Format")
        Menu.MenuItems(0).MenuItems.Add("&Font...", _
                                  AddressOf MenuFontOnClick)
        Menu.MenuItems(0).MenuItems.Add("&Background Color...", _
                                  AddressOf MenuColorOnClick)
    End Sub

    Sub MenuFontOnClick(ByVal obj As Object, ByVal ea As EventArgs)
        Dim fntdlg As New FontDialog()
        fntdlg.Font = Font
        fntdlg.Color = ForeColor
        fntdlg.ShowColor = True
        fntdlg.ShowApply = True
        AddHandler fntdlg.Apply, AddressOf FontDialogOnApply

        If fntdlg.ShowDialog() = DialogResult.OK Then
            Font = fntdlg.Font
            ForeColor = fntdlg.Color
            Invalidate()
        End If
    End Sub

    Sub MenuColorOnClick(ByVal obj As Object, ByVal ea As EventArgs)
        clrdlg.Color = BackColor

        If clrdlg.ShowDialog() = DialogResult.OK Then
            BackColor = clrdlg.Color
        End If
    End Sub

    Sub FontDialogOnApply(ByVal obj As Object, ByVal ea As EventArgs)
        Dim fntdlg As FontDialog = DirectCast(obj, FontDialog)
        Font = fntdlg.Font
        ForeColor = fntdlg.Color
        Invalidate()
    End Sub

    Protected Overrides Sub OnPaint(ByVal pea As PaintEventArgs)
        Dim grfx As Graphics = pea.Graphics
        grfx.DrawString("Hello common dialog boxes!", Font, _
                        New SolidBrush(ForeColor), 0, 0)
    End Sub
End Class

File: win-prog/ProgWinCS/Dialog Boxes/BetterFontAndColorDialogs/BetterFontAndColorDialogs.cs
//--------------------------------------------------------
// BetterFontAndColorDialogs.cs © 2001 by Charles Petzold
//--------------------------------------------------------
using System;
using System.Drawing;
using System.Windows.Forms;

class BetterFontAndColorDialogs:Form
{
     protected ColorDialog clrdlg = new ColorDialog();

     public static void Main()
     {
          Application.Run(new BetterFontAndColorDialogs());
     }
     public BetterFontAndColorDialogs()
     {
          Text = "Better Font and Color Dialogs";

          Menu = new MainMenu();
          Menu.MenuItems.Add("&Format");
          Menu.MenuItems[0].MenuItems.Add("&Font...",
                                        new EventHandler(MenuFontOnClick));
          Menu.MenuItems[0].MenuItems.Add("&Background Color...",
                                        new EventHandler(MenuColorOnClick));
     }
     void MenuFontOnClick(object obj, EventArgs ea)
     {
          FontDialog fontdlg = new FontDialog();

          fontdlg.Font  = Font;
          fontdlg.Color = ForeColor;
          fontdlg.ShowColor = true;
          fontdlg.ShowApply = true;
          fontdlg.Apply += new EventHandler(FontDialogOnApply);

          if(fontdlg.ShowDialog() == DialogResult.OK)
          {
               Font = fontdlg.Font;
               ForeColor = fontdlg.Color;
               Invalidate();
          }
     }
     void MenuColorOnClick(object obj, EventArgs ea)
     {
          clrdlg.Color = BackColor;

          if (clrdlg.ShowDialog() == DialogResult.OK)
               BackColor = clrdlg.Color;
     }
      void FontDialogOnApply(object obj, EventArgs ea)
     {
          FontDialog fontdlg = (FontDialog) obj;

          Font = fontdlg.Font;
          ForeColor = fontdlg.Color;
          Invalidate();
     }
    protected override void OnPaint(PaintEventArgs pea)
     {
          Graphics grfx = pea.Graphics;
          grfx.DrawString("Hello common dialog boxes!", Font, 
                          new SolidBrush(ForeColor), 0, 0);
     }
}